### PR TITLE
[CALCITE-6226] Wrong ISOWEEK and no ISOYEAR on BigQuery FORMAT_DATE

### PR DESCRIPTION
**Pull request by rorueda**
https://issues.apache.org/jira/browse/CALCITE-6226

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java [hunk 1]
```diff
@@ -132,17 +133,41 @@ public enum FormatElementEnum implements FormatElement {
       sb.append(String.format(Locale.ROOT, "%02d", calendar.get(Calendar.HOUR_OF_DAY)));
     }
   },
+  ID("u", "The weekday (Monday as the first day of the week) as a decimal number (1-7)") {
+    @Override public void format(StringBuilder sb, Date date) {
+      final Calendar calendar = Work.get().calendar;
+      calendar.setTime(date);
+      int weekDay = calendar.get(Calendar.DAY_OF_WEEK);
+      sb.append(String.format(Locale.ROOT, "%d", weekDay == 1 ? 7 : weekDay - 1));
```
Comment (tanclary) [inline core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:141]: Does this just convert from Sunday being 1 to Monday being 1?

Comment (rorueda) [reply core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:141]: Yes. Maybe change the comparison to `weekDay == Calendar.SUNDAY` to make it clearer? Do you see another way?

Comment (tanclary) [reply core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:141]: That could work, a simple comment would help also. It is no big deal either way :)

Comment (tanclary) [reply core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:141]: In my mind Calcite is complex enough that when we have opportunities to explain little pieces of logic to reduce mental load for the developer, we should

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:141]: The DOW function implementation assumes that Sunday = 1.
In Postgres DOW assumes that Sunday = 0.

File: core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java [hunk 2]
```diff
@@ -132,17 +133,41 @@ public enum FormatElementEnum implements FormatElement {
...
-      // TODO: ensure this is isoweek
-      final Calendar calendar = Work.get().calendar;
+      final Calendar calendar = Work.get().iso8601Calendar;
       calendar.setTime(date);
-      calendar.setFirstDayOfWeek(Calendar.MONDAY);
       sb.append(String.format(Locale.ROOT, "%02d", calendar.get(Calendar.WEEK_OF_YEAR)));
     }
   },
+  IYY("YY", "The ISO 8601 year without century as a decimal number. Each ISO year begins on "
+      + "the Monday before the first Thursday of the Gregorian calendar year.") {
```
[the remark this answers is not in the document]
Comment (rorueda) [reply core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:157]: Sure. Completely missed this test class.

File: core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java [hunk 3]
```diff
@@ -48,7 +48,8 @@ public enum FormatElementEnum implements FormatElement {
       sb.append(String.format(Locale.ROOT, "%2d", calendar.get(Calendar.YEAR) / 100 + 1));
     }
   },
-  D("F", "The weekday (Monday as the first day of the week) as a decimal number (1-7)") {
+  // TODO: Parsing of weekdays with sunday as first day of the week
```
[the remark this answers is not in the document]
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:51]: Please remove the TODO, log a jira case.

File: core/src/test/java/org/apache/calcite/util/format/FormatElementEnumTest.java [hunk 4]
```diff
@@ -77,8 +80,26 @@ class FormatElementEnumTest {
     assertFormatElement(FormatElementEnum.FF6, "2014-09-30T10:00:00.123456Z", "000123");
   }
 
-  @Test void testIW() {
-    assertFormatElement(FormatElementEnum.IW, "2014-09-30T10:00:00Z", "40");
+  @Test void testID() {
+    assertFormatElement(FormatElementEnum.ID, "2014-09-30T10:00:00Z", "2");
+  }
```
Comment (julianhyde) [inline core/src/test/java/org/apache/calcite/util/format/FormatElementEnumTest.java:85]: I don't understand why some of the tests in this class are parameterized and others are not.

Frankly I don't think the framework is helping make the tests more understandable. I would move to code, and add comments if a test is testing something subtle.

It seems to me that this test class should be very simple.

**Review by julianhyde — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rorueda) [reply core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:51]: Created CALCITE-6233 to handle the parsing of ISOWEEK.

It doesn't seem that parsing the weekday with Sunday as first day (1) is possible with java's DateFormat.

File: core/src/test/java/org/apache/calcite/util/format/FormatElementEnumTest.java (hunk 4 above)
Comment (rorueda) [reply core/src/test/java/org/apache/calcite/util/format/FormatElementEnumTest.java:85]: As there is one method for each format element, I didn't want to add new methods to test the edge cases of  ISOWEEK and ISOYEAR, and I thought a good solution was to make those tests parameterized.

But I agree that the way it was done didn't help in making the added test cases understandable. I removed the parameters and made more assertions inside the method (with a comment).

File: core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java (hunk 1 above)
Comment (rorueda) [reply core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:144]: Added a comment - don't think my suggestion would make it clearer.

Looking at Postgres documentation, the `EXTRACT(DOW...` assumes Sunday = 0, but the format functions `to_char, to_date,...` also assume Sunday = 1 for day of week.

File: core/src/test/java/org/apache/calcite/util/format/FormatElementEnumTest.java [hunk 5]
```diff
@@ -77,8 +77,32 @@ class FormatElementEnumTest {
...
 
-  @Test void testIW() {
+  @Test void testID() {
+    assertFormatElement(FormatElementEnum.ID, "2014-09-30T10:00:00Z", "2");
+  }
+
+  @Test
+  void testIW() {
     assertFormatElement(FormatElementEnum.IW, "2014-09-30T10:00:00Z", "40");
+    // edge case where ISO WEEK != WEEK
```
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/util/format/FormatElementEnumTest.java:87]: I have discovered that the Java Calendar class doesn't handle correctly some operations with dates before the Gregorian calendar introduction. Can you please add some tests for dates in that range?
https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-6252
Can you also confirm that the values have been validated against BigQuery?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 6]
```diff
@@ -12472,6 +12472,21 @@ void testTimestampDiff(boolean coercionEnabled) {
     f.checkScalar("FORMAT_DATE('The date is: %x', DATE '2008-12-25')",
         "The date is: 12/25/08",
         "VARCHAR(2000) NOT NULL");
+    f.checkScalar("FORMAT_DATE('%G-%V', DATE '2023-01-01')",
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:12475]: same comment about dates before the Gregorian calendar

File: core/src/test/java/org/apache/calcite/util/format/FormatElementEnumTest.java (hunk 5 above)
Comment (rorueda) [reply core/src/test/java/org/apache/calcite/util/format/FormatElementEnumTest.java]: I added additional test cases, all manually validated against BigQuery.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 6 above)
Comment (rorueda) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:12763]: Done. I also had to fix one test case for PostgreSQL ISOWEEK - also manually validated using dbfiddle.

**Comment by rorueda** [comment]
Since I had to rebase it onto master to address the last change request, I squashed the commits.

**Comment by julianhyde** [comment]
It always struck me as strange that this change is using Calendar. Day of the week is simple if you convert to Julian date. JulianOn Mar 9, 2024, at 07:59, rorueda ***@***.***> wrote:﻿
@rorueda commented on this pull request.



In core/src/test/java/org/apache/calcite/util/format/FormatElementEnumTest.java:
>      assertFormatElement(FormatElementEnum.IW, "2014-09-30T10:00:00Z", "40");
+    // edge case where ISO WEEK != WEEK

I added additional test cases, all manually validated against BigQuery.

**Review by mihaibudiu — APPROVED** [review]
I trust that the results were all compared against the reference databases, I didn't check myself.

**Review by jduo — APPROVED** [review]

**Comment by rorueda** [comment]
This is a bit old, so there were conflicts with the current master. I did a rebase and fixed the conflicts.

**Comment by rubenada** [comment]
LGTM

File: core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java [hunk 7]
```diff
@@ -436,8 +469,15 @@ static Work get() {
       return castNonNull(THREAD_WORK.get());
     }
 
-    final Calendar calendar =
-        Calendar.getInstance(DateTimeUtils.DEFAULT_ZONE, Locale.ROOT);
+    final Calendar calendar = new Calendar.Builder()
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:472]: According to this issue https://issues.apache.org/jira/browse/CALCITE-6580
setLocale should not be used.

Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:472]: That bug refers to `Locale.setDefault`, which is a global change (affects the whole JVM).

This change seems harmless, because the behavior is deterministic, and does not depend on the locale of the user. 

According to <a href="https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Calendar.Builder.html#setLocale(java.util.Locale)">setLocale<a> the locale will override the locale of the format element, which is a slight change in behavior, and if it causes a bug we can revisit.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 8]
```diff
@@ -4828,7 +4828,7 @@ void testBitGetFunc(SqlOperatorFixture f, String functionName) {
           "2",
           "VARCHAR NOT NULL");
       f.checkString("to_char(timestamp '2022-06-03 13:15:48.678', 'IW')",
-          "23",
+          "22",
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:4831]: this was a bug?

**Review by mihaibudiu — APPROVED** [review]

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 8 above)
Comment (rorueda) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:4831]: Yes, same ISOWEEK issue, but for PostgreSQL:  https://github.com/apache/calcite/pull/3653#discussion_r1518608626
